test/state/system_contracts.cpp

```diff
@@ -0,0 +1,84 @@
+// evmone: Fast Ethereum Virtual Machine implementation
+// Copyright 2023 The evmone Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "system_contracts.hpp"
+#include "host.hpp"
+#include "state.hpp"
+
+namespace evmone::state
+{
+namespace
+{
+/// The address of the sender of the system calls (EIP-4788).
+constexpr auto SYSTEM_ADDRESS = 0xfffffffffffffffffffffffffffffffffffffffe_address;
+
+/// The address of the system contract storing the root hashes of beacon chain blocks (EIP-4788).
+constexpr auto BEACON_ROOTS_ADDRESS = 0x000F3df6D732807Ef1319fB7B8bB8522d0Beac02_address;
+
+constexpr auto HISTORY_STORAGE_ADDRESS = 0x0aae40965e6800cd9b1f4b05ff21581047e3f91e_address;
+
+/// Information about a registered system contract.
+struct SystemContract
+{
+    using GetInputFn = bytes_view(const BlockInfo&) noexcept;
+
+    evmc_revision since = EVMC_MAX_REVISION;  ///< EVM revision in which added.
+    address addr;                             ///< Address of the system contract.
+    GetInputFn* get_input = nullptr;          ///< How to get the input for the system call.
+};
+
+/// Registered system contracts.
+constexpr std::array SYSTEM_CONTRACTS{
+    SystemContract{EVMC_CANCUN, BEACON_ROOTS_ADDRESS,
+        [](const BlockInfo& block) noexcept { return bytes_view{block.parent_beacon_block_root}; }},
+    SystemContract{EVMC_PRAGUE, HISTORY_STORAGE_ADDRESS,
+        [](const BlockInfo& block) noexcept {
+            return bytes_view{block.known_block_hashes.at(block.number - 1)};
+        }},
+};
+
+static_assert(std::ranges::is_sorted(SYSTEM_CONTRACTS,
+                  [](const auto& a, const auto& b) noexcept { return a.since < b.since; }),
+    "system contract entries must be ordered by revision");
+
+}  // namespace
+
+void system_call(State& state, const BlockInfo& block, evmc_revision rev, evmc::VM& vm)
+{
+    for (const auto& [since, addr, get_input] : SYSTEM_CONTRACTS)
+    {
+        if (rev < since)
+            return;  // Because entries are ordered, there are no other contracts for this revision.
+
+        const auto acc = state.find(addr);
+        if (acc == nullptr)
+            continue;  // Nothing to do if the contract doesn't exist.
```

Comment:
in what scenario does this happen? sounds more like an assert condition, not a silent `continue`